### DISCUSSION TITLE
How to pass labels as a function argument?

### DISCUSSION THREAD

**Discussion by jaskij** [metrics-rs/metrics#349] [Q&A] [answered]
I am working on genericizing some code commonly used between several of my microservices. This raised some issues, one of them being able to pass a metric's labels to a function.

After looking through the API, particularly at [`impl From for Label`](https://docs.rs/metrics/latest/metrics/struct.Label.html#impl-From%3C%26(K%2C%20V)%3E-for-Label), it seems that the best way is to accept an `Option<&[(SharedString, SharedString)]`, and somewhere inside the macro it is then coerced into correct values. Is this right?

Full code for the function:

```rust

    pub(crate) fn new_with_metric_name(
        db: Database,
        engine: EngineId,
        metric_name: &'static str,
        labels: Option<&[(SharedString, SharedString)]>,
    ) -> Self {
        let histogram = match labels {
            None => register_histogram!(metric_name),
            Some(labels) => register_histogram!(metric_name, labels),
        };
        describe_histogram!(
            metric_name,
            metrics::Unit::Seconds,
            "Time taken to insert data into the database"
        );
        Inserter { db, engine, histogram, _message_phantom: Default::default() }
    }
```

**Comment by tobz** [comment, marked as answer]
That's certainly one way to do it, yeah.

In this case, based on the given construction, both the metric name and labels are considered "non static", so it will end up calling `metrics::Key::from_parts(metric_name, labels)` to construct the key. [That method](https://docs.rs/metrics/latest/metrics/struct.Key.html#method.from_parts) depends on [IntoLabels](https://docs.rs/metrics/latest/metrics/trait.IntoLabels.html), which you could also use here to generalize your function, if you wanted to do so:

```rust
pub(crate) fn new_with_metric_name<L>(
    db: Database,
    engine: EngineId,
    metric_name: &'static str,
    labels: Option<L>,
) -> Self
where
  L: metrics::IntoLabels,
{
}
```

**Reply by jaskij** [reply]
Thank you for the info! This does save me a lot of digging through that proc_macro code.

That's a nice generalization, thank you for that, but it is a pain to call with `labels = None`, so I'll probably need to add another function for the no-labels case.
